### PR TITLE
Implements parseQuery in uri module.

### PR DESCRIPTION
**Pull request by dom96**

### CODE REVIEW

File: lib/pure/uri.nim [hunk 1]
```diff
@@ -380,6 +380,57 @@ proc `$`*(u: Uri): string =
...
+
+proc `[]`*(query: Query, key: string): QueryValues =
+  return Table[string, seq[string]](query)[key].QueryValues
+
+proc `[]`*(query: Query, key: string, i: int): string =
+  return Table[string, seq[string]](query)[key][i]
+
+proc `[]`*(values: QueryValues, i: int): string = seq[string](values)[i]
+
+converter toString*(value: QueryValues): string =
```
Comment (Araq) [inline lib/pure/uri.nim]: I don't like converters.

File: lib/pure/uri.nim [hunk 2]
```diff
@@ -380,6 +380,57 @@ proc `$`*(u: Uri): string =
     result.add("#")
     result.add(u.anchor)
 
+type
+  Query* = distinct Table[string, seq[string]]
```
Comment (Araq) [inline lib/pure/uri.nim]: A table can easily hold multiple values for the same key, no ``seq[string]`` required.

File: lib/pure/uri.nim [hunk 3]
```diff
@@ -380,6 +380,57 @@ proc `$`*(u: Uri): string =
...
+  return Table[string, seq[string]](query)[key][i]
+
+proc `[]`*(values: QueryValues, i: int): string = seq[string](values)[i]
+
+converter toString*(value: QueryValues): string =
+  ## Implicit converter which retrieves the first value in a list
+  ## of query values.
+  return seq[string](value)[0]
+
+proc parseQuery*(query: string): Query =
```
Comment (Araq) [inline lib/pure/uri.nim]: This should be an iterator that yields the (key, value) pairs IMO. Not yet another API that encourages bloat everywhere.

**Comment by dom96** [comment]
This implementation mirrors that of ``HttpHeaders``, should this be changed too?

**Comment by Araq** [comment]
> This implementation mirrors that of HttpHeaders, should this be changed too?

If possible. :-)

**Comment by dom96** [comment]
So what's the alternative?

**Comment by Araq** [comment]
```nim

proc parseQuery*(query: string): OrderedTable[string, string]
```

I would like much better.

**Comment by dom96** [comment]
That wouldn't work for parsing "q=foo&q=bar"

**Comment by Araq** [comment]
It would, an OrderedTable can save multiple values under the same key.

**Comment by dom96** [comment]
How do I retrieve the multiple values?

**Comment by kennymalac** [comment]
@dom96  Can we get this in? Noticed this was missing!

**Comment by dom96** [comment]
@kennymalac maybe you'd like to push this over the line? :) My time is spread pretty thin across far too many projects these days

**Comment by Araq** [comment]
Stalled PR, closing.

**Comment by mildred** [comment]
I believe this should work:

```diff
diff -u /home/mildred/.choosenim/toolchains/nim-1.0.6/lib/pure/uri.nim.old /home/mildred/.choosenim/toolchains/nim-1.0.6/lib/pure/uri.nim
--- /home/mildred/.choosenim/toolchains/nim-1.0.6/lib/pure/uri.nim.old	2020-04-20 22:34:31.598977025 +0200
+++ /home/mildred/.choosenim/toolchains/nim-1.0.6/lib/pure/uri.nim	2020-04-20 23:12:00.996389127 +0200
@@ -37,7 +37,7 @@
 ##    else:
 ##      echo "Wrong format"
 
-import strutils, parseutils
+import strutils, parseutils, tables
 type
   Url* = distinct string
 
@@ -148,6 +148,33 @@
       result.add('=')
       result.add(encodeUrl(val, usePlus))
 
+proc parseQuery*(query: string): OrderedTable[string, string] =
+  ## Parses a URL query containing key-value pairs delimited by
+  ## ampersands and URL-decode the individual key and values.
+  runnableExamples:
+    doAssert parseQuery("q=Nim+programming+language&lang=en")["q"] ==
+                        "Nim programming language"
+
+    let u = parseUri("http://google.com/?q=hello+world")
+    doAssert parseQuery(u.query)["q"] == "hello world"
+
+  var res = initOrderedTable[string, string]()
+
+  var i = 0
+  while i < query.len:
+    var key = ""
+    var value = ""
+    i += query.parseUntil(key, {'=', '&'}, i)
+    if query[i] == '=':
+      i.inc() # Skip =
+
+    i += query.parseUntil(value, {'&'}, i)
+    if i < query.len and query[i] == '&': i.inc()
+
+    res.add(decodeUrl(key), decodeUrl(value))
+
+  return res
+
 proc parseAuthority(authority: string, result: var Uri) =
   var i = 0
   var inPort = false
@@ -735,4 +762,10 @@
       var foo1 = parseUri("http://example.com/foo?do=do&bar")
       doAssert foo == foo1
 
+  block: # parseQuery
+    doAssert "foo=bar".parseQuery()["foo"] == "bar"
+    doAssert "foo=bar&alpha=beta".parseQuery()["foo"] == "bar"
+    doAssert "foo=bar&alpha=beta".parseQuery()["alpha"] == "beta"
+    doAssert "foo=&alpha=beta".parseQuery().pairs["foo"] == ""
+
   echo("All good!")
```

**Comment by dom96** [comment]
Consider making a PR.

**Comment by mildred** [comment]
Yep, I'm just new to Nim so I'll need to get the setup working properly.

And since then, I discovered cgi.decodeData which does already the right 
thing (and seems more robust because 
https://play.nim-lang.org/#ix=2jad). The question is: do we duplicate 
identical code? Do we reference one in the other?

**Comment by timotheecour** [comment]
> The question is: do we duplicate identical code? Do we reference one in the other?

@mildred 
cgi already imports uri so we should move the implementation to uri and re-export the symbol in cgi

**Comment by ringabout** [comment]
Now, you can use [uri.decodeQuery](https://github.com/nim-lang/Nim/blob/146beb2797982e2b80ab83c227b7279cad4e0943/lib/pure/uri.nim#L156).

**Pull request closed without merging**